File: engine/nexgen_engine/core/schema_versions.py
"""Zentrale Komptabilitaets-Matrix fuer Skill-Schema-Versionen.

Hintergrund: Der Skill wird per `git pull` aktualisiert, Projekte aber
bleiben in ihrem Zustand. Wenn ein Skill mit einer Projekt-Schema-Version
konfrontiert wird, die er nicht kennt, riskieren wir stille Datenkorruption
(neue Pflichtfelder fehlen, alte Felder werden falsch interpretiert).

Drei Faelle:
- **Projekt-Version <= Skill-Version**: Skill kennt das Schema, kann lesen.
  Wenn niedriger → Migration moeglich (siehe `<modul>.migrate`).
- **Projekt-Version == Skill-Version**: alles in Ordnung.
- **Projekt-Version > Skill-Version**: HART STOPPEN. Skill verweigert
  Weiterarbeit, User soll Skill aktualisieren (git pull + CLAUDE.md neu
  einlesen).

Dieses Modul ist die Single Source of Truth fuer „was der aktuelle Skill
kennt". Wer ein Schema bumpen will, traegt es hier ein UND schreibt die
Migration.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Literal

import yaml

# ---------------------------------------------------------------------------
# Komptabilitaets-Matrix
# ---------------------------------------------------------------------------
#
# Pro Schema:
#   `current`     — Was der Skill HEUTE schreibt.
#   `supported`   — Was der Skill (per Migration / Tolerant-Reader) lesen kann.
#                    `current` ist immer in `supported`.
#
# Versionierungs-Pattern: `<name>/v<N>` (z.B. `bible/v5`).

SCHEMA_MATRIX: dict[str, dict[str, object]] = {
    "bible": {
        "current": "bible/v5",
        "supported": ("bible/v4", "bible/v5"),
    },
    "shotlist": {
        "current": "shotlist/v3",
        "supported": ("shotlist/v1", "shotlist/v2", "shotlist/v3"),
    },
    "brief": {
        "current": "brief/v1",
        "supported": ("brief/v1",),
    },
    "frame_audit": {
        "current": "frame_audit/v1",
        "supported": ("frame_audit/v1",),
    },
    # Treatment hat kein versioniertes schema-Feld (Markdown + Frontmatter),
    # daher nicht in der Matrix.
    # Storyboard nutzt YAML, Schema-String `storyboard/v1` — versionsstabil
    # bis substanziell verändert.
    "storyboard": {
        "current": "storyboard/v1",
        "supported": ("storyboard/v1",),
    },
}


VersionStatus = Literal["match", "behind", "ahead", "unknown", "missing"]


@dataclass(frozen=True)
class VersionFinding:
    """Befund pro Projekt-Artefakt vs. Skill-Matrix."""
    artifact: str
    """z.B. 'bible.yaml', 'shotlist/current.yaml', 'brief.yaml'."""
    schema_field: str
    """Schluessel in SCHEMA_MATRIX, z.B. 'bible'."""
    project_version: str | None
    """Version-String aus dem Projekt-File, oder None wenn File fehlt."""
    skill_current: str
    """Was der Skill heute schreibt."""
    status: VersionStatus
    """match | behind (Migration moeglich) | ahead (Skill veraltet, HART STOP) |
    unknown (Version-String unbekannt) | missing (File existiert nicht)."""
    message: str


def _parse_v(s: str) -> int | None:
    """`bible/v5` -> 5. None wenn nicht parsebar."""
    if not s or "/v" not in s:
        return None
    try:
        return int(s.rsplit("/v", 1)[1])
    except ValueError:
        return None


def _classify(project_v: str | None, schema_key: str) -> tuple[VersionStatus, str]:
    info = SCHEMA_MATRIX[schema_key]
    current = info["current"]  # type: ignore[index]
    supported = info["supported"]  # type: ignore[index]
    if project_v is None:
        return "missing", f"File fehlt — Skill schreibt '{current}'."
    if project_v == current:
        return "match", f"OK — beide auf '{current}'."
    if project_v in supported:
        return "behind", (
            f"Projekt auf '{project_v}', Skill aktuell '{current}'. "
            f"Migration empfohlen (siehe `<modul> migrate`)."
        )
    # Nicht supported. Numerisch vergleichen, um 'ahead' vs 'unknown' zu trennen.
    p = _parse_v(project_v)
    c = _parse_v(current)  # type: ignore[arg-type]
    if p is not None and c is not None and p > c:
        return "ahead", (
            f"Projekt auf '{project_v}', Skill kennt nur bis '{current}'. "
            "Skill ist veraltet — HART STOP. "
            "Aktualisiere den Skill (`git pull` im Skill-Repo) und lies "
            "CLAUDE.md neu, bevor Du an diesem Projekt weiterarbeitest."
        )
    return "unknown", (
        f"Projekt deklariert '{project_v}', aber das ist weder current "
        f"('{current}') noch in supported {supported}. "
        "Manuelle Klärung nötig."
    )


def _read_schema_field(path: Path) -> str | None:
    if not path.exists():
        return None
    try:
        data = yaml.safe_load(path.read_text(encoding="utf-8"))
    except (yaml.YAMLError, OSError):
        return None
    if not isinstance(data, dict):
        return None
    v = data.get("schema")
    return v if isinstance(v, str) else None


# Mapping von Projekt-Dateipfad → Schema-Key
_ARTIFACTS: tuple[tuple[str, str], ...] = (
    ("bible/bible.yaml", "bible"),
    ("shotlist/current.yaml", "shotlist"),
    ("brief.yaml", "brief"),
    ("storyboard/current.yaml", "storyboard"),
)


def check_project_versions(project_dir: Path) -> list[VersionFinding]:
    """Pruefe alle bekannten Projekt-Artefakte gegen die Skill-Matrix.

    Findet keine `frame_audit/*.yaml`-Files automatisch — die werden
    pro Render geschrieben und entstehen erst dynamisch. Ihr Schema
    bleibt bislang v1.
    """
    findings: list[VersionFinding] = []
    for rel_path, schema_key in _ARTIFACTS:
        path = project_dir / rel_path
        proj_v = _read_schema_field(path)
        status, msg = _classify(proj_v, schema_key)
        findings.append(VersionFinding(
            artifact=rel_path,
            schema_field=schema_key,
            project_version=proj_v,
            skill_current=SCHEMA_MATRIX[schema_key]["current"],  # type: ignore[arg-type]
            status=status,
            message=msg,
        ))
    return findings


def any_ahead(findings: list[VersionFinding]) -> bool:
    """Mindestens ein Artefakt deklariert eine Version, die der Skill nicht
    kennt — Phase-Code soll dann hart abbrechen."""
    return any(f.status == "ahead" for f in findings)

File: Engine/Sources/NexGenEngine/Artifacts/SchemaVersions.swift
import Foundation
import Yams

/// Central schema-version compatibility matrix. Port of
/// `core/schema_versions.py`.
///
/// Background: the skill is updated via `git pull`, but projects stay in
/// whatever state they were saved. If a skill encounters a project schema
/// version it doesn't know, silent data corruption is possible (new
/// mandatory fields missing, old fields misinterpreted). Three cases:
/// - project version <= skill version: skill knows the schema, can read
///   (migration possible if lower).
/// - project version == skill version: fine.
/// - project version > skill version: HARD STOP — skill must be updated.
public enum SchemaVersions {
    public struct Info: Sendable {
        public let current: String
        public let supported: [String]
    }

    /// Per schema: `current` is what the skill writes today; `supported` is
    /// what it can read (via migration / tolerant reader). `current` is
    /// always included in `supported`.
    public static let matrix: [String: Info] = [
        "bible": Info(current: "bible/v5", supported: ["bible/v4", "bible/v5"]),
        "shotlist": Info(current: "shotlist/v3", supported: ["shotlist/v1", "shotlist/v2", "shotlist/v3"]),
        "brief": Info(current: "brief/v1", supported: ["brief/v1"]),
        "ledger": Info(current: "ledger/v1", supported: ["ledger/v1"]),
        "frame_audit": Info(current: "frame_audit/v1", supported: ["frame_audit/v1"]),
        // Treatment has no versioned `schema` field (markdown + frontmatter),
        // so it is not in the matrix.
        "storyboard": Info(current: "storyboard/v1", supported: ["storyboard/v1"]),
    ]

    public enum Status: String, Sendable, Equatable {
        case match
        case behind
        case ahead
        case unknown
        case missing
    }

    public struct Finding: Sendable, Equatable {
        /// e.g. "bible.yaml", "shotlist/current.yaml", "brief.yaml".
        public let artifact: String
        /// Key in `matrix`, e.g. "bible".
        public let schemaField: String
        /// Version string from the project file, or nil if the file is missing.
        public let projectVersion: String?
        /// What the skill writes today.
        public let skillCurrent: String
        public let status: Status
        public let message: String
    }

    /// `"bible/v5"` -> `5`. `nil` if unparsable.
    static func parseVersion(_ s: String) -> Int? {
        guard !s.isEmpty, let range = s.range(of: "/v") else { return nil }
        return Int(s[range.upperBound...])
    }

    static func classify(projectVersion: String?, schemaKey: String) -> (Status, String) {
        guard let info = matrix[schemaKey] else {
            // Not reachable from `checkProjectVersions` (keys are fixed), but
            // keep a safe fallback for direct callers.
            return (.unknown, "unknown schema key \(schemaKey)")
        }
        let current = info.current
        let supported = info.supported
        guard let projectVersion else {
            return (.missing, "File fehlt — Skill schreibt '\(current)'.")
        }
        if projectVersion == current {
            return (.match, "OK — beide auf '\(current)'.")
        }
        if supported.contains(projectVersion) {
            return (
                .behind,
                "Projekt auf '\(projectVersion)', Skill aktuell '\(current)'. "
                    + "Migration empfohlen (siehe `<modul> migrate`)."
            )
        }
        let p = parseVersion(projectVersion)
        let c = parseVersion(current)
        if let p, let c, p > c {
            return (
                .ahead,
                "Projekt auf '\(projectVersion)', Skill kennt nur bis '\(current)'. "
                    + "Skill ist veraltet — HART STOP. "
                    + "Aktualisiere den Skill (`git pull` im Skill-Repo) und lies "
                    + "CLAUDE.md neu, bevor Du an diesem Projekt weiterarbeitest."
            )
        }
        return (
            .unknown,
            "Projekt deklariert '\(projectVersion)', aber das ist weder current "
                + "('\(current)') noch in supported \(supported). "
                + "Manuelle Klärung nötig."
        )
    }

    static func readSchemaField(at url: URL) -> String? {
        guard let text = try? String(contentsOf: url, encoding: .utf8),
              let node = try? Yams.compose(yaml: text),
              case .mapping(let mapping) = node,
              let value = mapping["schema"],
              case .scalar(let scalar) = value
        else { return nil }
        return scalar.string
    }

    /// Mapping of project file path (relative to the data root) -> schema key.
    /// Does not find `frame_audit/*.yaml` files automatically — those are
    /// written per-render and only appear dynamically; their schema stays v1.
    static let artifacts: [(path: String, schemaKey: String)] = [
        ("bible/bible.yaml", "bible"),
        ("shotlist/current.yaml", "shotlist"),
        ("brief.yaml", "brief"),
        ("storyboard/current.yaml", "storyboard"),
    ]

    /// Check all known project artifacts against the skill matrix.
    public static func checkProjectVersions(dataRoot: URL) -> [Finding] {
        artifacts.map { entry in
            let url = dataRoot.appendingPathComponent(entry.path)
            let projectVersion = readSchemaField(at: url)
            let (status, message) = classify(projectVersion: projectVersion, schemaKey: entry.schemaKey)
            return Finding(
                artifact: entry.path,
                schemaField: entry.schemaKey,
                projectVersion: projectVersion,
                skillCurrent: matrix[entry.schemaKey]!.current,
                status: status,
                message: message
            )
        }
    }

    /// At least one artifact declares a version the skill doesn't know —
    /// phase code should hard-stop.
    public static func anyAhead(_ findings: [Finding]) -> Bool {
        findings.contains { $0.status == .ahead }
    }
}
